### PR TITLE
feat: prism support for syntax highlighting and copy button on code blocks

### LINKED ISSUE #3049: add a copy button to code examples
**This is a Feature Request**

**What would you like to be added**

Code examples in tutorials, READme files, and guides could use a copy button so people don't miss a single thing when they copy code.

**Why is this needed**
Better user experience.

**Comments**

Once this PR, https://github.com/google/docsy/pull/325, is merged Jenkins X docs should be able to add copy button to code examples quite easily. 🤞

### PR DESCRIPTION
**Pull request by borntorock**
Resolves #3049 

Signed-off-by: borntorock <borntorock@users.noreply.github.com>

### CODE REVIEW

**Comment by ankitm123** [comment]
I think you also need to pull the latest docsy master (included as a submodule here). I dont see any copy to clipboard in the preview. Also most probably by default terraform syntax highlighting may not be enabled, so u have to add the required css/js files.

**Comment by borntorock** [comment]
> I think you also need to pull the latest docsy master (included as a submodule here). I dont see any copy to clipboard in the preview. Also most probably by default terraform syntax highlighting may not be enabled, so u have to add the required css/js files.

Yes, this PR does not add copy as of now and still in progress. Thanks for pointing about updating the docsy submodule. I'm not sure which terraform syntax highlighting are you referring to? Can you please clarify.

**Comment by ankitm123** [comment]
For a WIP PR, you can add WIP to the title of the PR :)
Terraform codeblocks, for example: https://jenkins-x.io/docs/install-setup/create-cluster/eks/#cluster-provisioning

**Comment by borntorock** [comment]
Updating the docsy theme submodule without enabling the `prism_syntax_highlighting` is making our code blocks blurred.

@ankitm123 Any idea / suggestions that might help resolve this. Looks like the updated javascript and css code in the docsy theme is overwriting the font of the code blocks, but not sure which one.

**Comment by MarckK** [comment]
Looks good, @borntorock. Not blurry for me. 👍

**Comment by borntorock** [comment]
> Looks good, @borntorock. Not blurry for me.

Yes, you're right. It was my browser's cache, displaying the blurred code blocks. Let me know check again after enabling the `prism_syntax_highlighting`.

**Comment by borntorock** [comment]
the last build in this PR has the copy button on the code blocks, including terraform code blocks. Only thing is we have blurred code blocks but if we hover and preview the site in moon mode (black background), then the text in the code blocks is clearly visible. 

Looks due to some CSS over-ridden, but couldn't figure out which part of CSS is making this happen.

File: layouts/partials/head-css.html [hunk 1]
```diff
@@ -12,3 +12,7 @@
 <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/docsearch.js@2/dist/cdn/docsearch.min.css" />
 <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.15.10/styles/monokai.min.css">
 <link rel="stylesheet" href='{{ "css/style.css" | absURL }}' />
+{{ with .Site.Params.prism_syntax_highlighting }}
```
Comment (ankitm123) [inline layouts/partials/head-css.html:15]: You can remove this, as this is already included in the docsy folder.

File: layouts/partials/scripts.html [hunk 2]
```diff
@@ -83,4 +83,8 @@
 	}
 
 </script>
-{{ partial "hooks/body-end.html" . }}
\ No newline at end of file
+{{ with .Site.Params.prism_syntax_highlighting }}
```
Comment (ankitm123) [inline layouts/partials/scripts.html:86]: Same, remove this as well. Don't need to include the prism css and js files

**Comment by ankitm123** [comment]
@borntorock Left 2 comments.
The reason for the blurry text is a combination of these 2 lines: https://github.com/google/docsy/blob/master/static/css/prism.css#L11 and https://github.com/google/docsy/blob/master/static/css/prism.css#L13.

So, you can just copy the `prism.css` file into the static/css folder, and remove those 2 lines, and it will render properly.

Another thing, it would be nice if you can change the language from `javascript` to `terraform` [here](https://github.com/jenkins-x/jx-docs/blob/master/content/en/docs/_index.html#L56) and [here](https://github.com/jenkins-x/jx-docs/blob/master/content/en/docs/_index.html#L92).
Switching to prism from chroma makes them look bad (and also it's terraform and not js :D)

File: layouts/partials/scripts.html (hunk 2 above)
Comment (borntorock) [reply layouts/partials/scripts.html:86]: We do need to include the prism css and js files, as the prism js file is referenced in `layouts/partials/scripts.html` which is actually being over-ridden with the `scripts.html` we have in our jx-docs repository. Then, there's no reference to the prism js file. 

I did cross checked (after your comment) to see if it works or not but it didn't, by excluding the prism js file here.

**Comment by borntorock** [comment]
Thanks for pointing out the suggestions. It does resolves the blurry text now. Also, we do need to reference the prism css and js files in `head-css.html` and `scripts.html` resp., more details [here](https://github.com/jenkins-x/jx-docs/pull/3057#discussion_r490723615).

> Another thing, it would be nice if you can change the language from `javascript` to `terraform` [here](https://github.com/jenkins-x/jx-docs/blob/master/content/en/docs/_index.html#L56) and [here](https://github.com/jenkins-x/jx-docs/blob/master/content/en/docs/_index.html#L92).

Done!

File: layouts/partials/head-css.html (hunk 1 above)
Comment (borntorock) [reply layouts/partials/head-css.html:15]: the prism css file also needs to be referenced but now from our static css folder similar to style.css.

**Comment by borntorock** [comment]
@ankitm123 @MarckK the preview environment looks better with the clear text and copy button hover on the code blocks, so good to rebase and merge this PR. 

Let me know, if there's still something to add or change in this MR.

**Comment by MarckK** [comment]
Nice work, @borntorock!

The only oddity with the code snippets is that some characters have grey highlighting. For example, [here](http://jx-docs.jx-jenkins-x-jx-docs-pr-3057.jenkins-x.live/docs/v3/guides/operator/):

Screenshot 2020-09-18 at 09 18 27

and [here](http://jx-docs.jx-jenkins-x-jx-docs-pr-3057.jenkins-x.live/docs/):
Screenshot 2020-09-18 at 09 21 53

**Comment by ankitm123** [comment]
@MarckK that happens because we use the default theme. @borntorock I think if you use the prism css (I think the js one remains the same) file from the `prism-tomorrow` theme (This is the theme I use for my personal blog), it will fix the issue: https://prismjs.com/download.html#themes=prism-tomorrow&languages=markup+css+clike+javascript+bash+css-extras+docker+go+groovy+makefile+yaml&plugins=toolbar+copy-to-clipboard

**Comment by ankitm123** [comment]
I think it looks good now @borntorock @MarckK. Time to get it merged? Any other issues?

**Pull request merged**